SUBMARINE-1023. Submarine SDK sqlalchemy store (model registry)

KUAN-HSUN-LI opened the pull request:
### What is this PR for?
* Implement the model registry SQL method in Python SDK
* Apply sqlalchemy mypy checks
* Replace submarine tracking_uri with db_uri

### What type of PR is it?
[Feature]

### Todos


### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-1023

### How should this be tested?
All of the tests are provided in `submarine-sdk/pysubmarine/tests/store/model_registry/test_sqlalchemy_store.py`

### Screenshots (if appropriate)

### Questions:
* Do the license files need updating? No
* Are there breaking changes for older versions? No
* Does this need new documentation? No

KUAN-HSUN-LI commented:
@pingsutw @jeff-901  Can you help me review the code? Thanks

jeff-901 commented on submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:
```diff
@@ -0,0 +1,570 @@
+            sql_model_version = self._get_model_version(session, name, version)
+            sql_model_version.description = description
+            sql_model_version.last_updated_time = update_time
+            self._save_to_db(session, sql_model_version)
```
Is it necessary to check the length of description?

KUAN-HSUN-LI replied:
I will add the check

jeff-901 commented on submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:
```diff
@@ -0,0 +1,570 @@
+        conditions = [SqlModelVersion.name == name]
+        if filter_tags is not None:
+            conditions = [
+                SqlModelVersion.model_tags.any(SqlModelTag.tag.contains(tag)) for tag in filter_tags
```
Is it wrong to reassign "conditions" when filter_tags is not None?

KUAN-HSUN-LI replied:
Thanks for point out this problem

jeff-901 commented on submarine-sdk/pysubmarine/tests/store/model_registry/test_sqlalchemy_store.py:
```diff
@@ -0,0 +1,731 @@
+        # empty result
+        results = self.store.list_registered_model(filter_tags=tags + other_tag)
+        self.assertEqual(len(results), 0)
+        self.assertEqual(results, [])
```
This test only contains one registered model name. Adding other registered model name helps to test list_model_version thoroughly when both name and tags are provided.

pingsutw commented on submarine-sdk/pysubmarine/submarine/utils/validation.py:
```diff
@@ -116,6 +117,45 @@ def validate_param(key, value):
+    if tag is None or tag == "":
+        raise SubmarineException("Tag cannot be empty.")
+    if not _VALID_PARAM_AND_METRIC_NAMES.match(tag):
+        raise SubmarineException("Invalid tag name: '%s'. %s" % (tag, _BAD_CHARACTERS_MESSAGE))
```
```suggestion
        raise SubmarineException(f"Invalid tag name: {tag}. {_BAD_CHARACTERS_MESSAGE}")
```

KUAN-HSUN-LI commented on submarine-sdk/pysubmarine/github-actions/test-requirements.txt:
```diff
@@ -23,11 +23,12 @@ pytest==3.2.1
 pytest-localserver==0.5.0
 pylint==2.5.2
-sqlalchemy==1.3.0
+sqlalchemy >= 1.4.0
```
Sure, I will remove it.

KUAN-HSUN-LI commented on submarine-sdk/pysubmarine/submarine/store/model_registry/abstract_store.py:
```diff
@@ -0,0 +1,234 @@
+        pass
+
+    @abstractmethod
+    def create_registered_model(
```
model registry including `registered_model`, `registered_model_tag`, `model_version`, `model_version_tag` four tables.
tracking including `param`, `metric`, `experiment` three tables.
In my opinion, it is more clear to put these functions in different directories. In this PR, I only worked on the model registry directory maybe I should also refactor the tracking directory and implement the `experiment` table.

KUAN-HSUN-LI commented on submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:
```diff
@@ -0,0 +1,572 @@
+        experiment_id: str,
+        dataset: str = None,
+        description: str = None,
+        tags: List[str] = None,
```
`registered model tag` focuses on what is the goal of this model or the category of this model. `model version tag` focuses on what this model perform or feature of this model. I think it is a little confused in these two parts. Hopefully, this explanation can help.

KUAN-HSUN-LI commented:
@pingsutw @jeff-901 Thanks for your review. I have replaced `registered model` with `model container` and replaced `model version` with `model metadata`. Additionally, I have fixed serval bugs.

jeff-901 commented on submarine-sdk/pysubmarine/submarine/utils/validation.py:
```diff
@@ -116,8 +117,48 @@ def validate_param(key, value):
+        raise SubmarineException("Registered model name cannot be empty.")
+
+
+def validate_model_version(model_version: int) -> None:
```
model_version forget to change to model metadata

KUAN-HSUN-LI replied:
I think this function name is fine.

jeff-901 replied:
ok

jeff-901 commented on submarine-sdk/pysubmarine/tests/entities/model_registry/test_model_version.py:
```diff
@@ -15,12 +15,11 @@
 
 from datetime import datetime
```
This file name should be changed to test_model_metadata.py

jeff-901 reviewed (approved):
LGTM

KUAN-HSUN-LI commented:
@pingsutw @jeff-901 I am so sorry for changing the table name in the database again. I think the previous names are the best. I will explain it as follow:
workflow of model registry:
1. save models in s3 bucket (unregistered model)
2. choose the best model and registry it (registered model, version 1)
3. train the model with different parameters or datasets and save models in s3 bucket (unregistered model)
4. Again, choose the best model and registry it (registered model, version 2)

Summary:
* The table is storing registered models so the name `registered_model` is more proper.
* The version of the registered model with the same name will define a different version which will be stored in another table. It is also proper to name the table `model_version`

We also need two tag tables.
example:
![image](https://user-images.githubusercontent.com/38066413/134780767-b80258fa-a901-4dd8-a693-0f9d69c24dfa.png)
